src/menu/menu.c

```diff
@@ -856,22 +970,203 @@ menu_open(struct menu *menu, int x, int y)
+	parse_buf(ctx->server, &ctx->buf);
+	current_menu = current_menu->parent;
+	ctx->item->submenu = menu_get_by_id(ctx->server, ctx->item->id);
+	menu_level--;
+
+	if (!ctx->item->submenu) {
+		goto out;
+	}
+
+	/*
+	 * TODO: refactor validate() and post_processing() to only
+	 * operate from current point onwards
+	 */
+
+	/* Set menu-widths before configuring */
+	post_processing(ctx->server);
+
+	/*
+	 * TODO:
+	 * (1) Combine this with the code in get_submenu_position()
+	 *     and/or menu_configure()
+	 * (2) Take into account menu_overlap_{x,y}
+	 */
+	enum menu_align align = ctx->item->parent->align;
+	int x = ctx->item->parent->scene_tree->node.x;
+	int y = ctx->item->parent->scene_tree->node.y + ctx->item->tree->node.y;
+	if (align & LAB_MENU_OPEN_RIGHT) {
+		x += ctx->item->parent->size.width;
+	}
+	menu_configure(ctx->item->submenu, x, y, align);
+
+	validate(ctx->server);
+
+	/* Sync the triggering view */
+	ctx->item->submenu->triggered_by_view = ctx->item->parent->triggered_by_view;
+	/* Ensure the submenu has its parent set correctly */
+	ctx->item->submenu->parent = ctx->item->parent;
+	/* And open the new submenu tree */
+	wlr_scene_node_set_enabled(
+		&ctx->item->submenu->scene_tree->node, true);
+out:
+	ctx->item->parent->selection.menu = ctx->item->submenu;
+
+clean_up:
+	wl_event_source_remove(ctx->event_read);
+	wl_event_source_remove(ctx->event_timeout);
+	spawn_piped_close(ctx->pid, ctx->pipe_fd);
+	free(ctx->buf.buf);
+	free(ctx);
+	return 0;
+}
+
+static void
+parse_pipemenu(struct menuitem *item)
+{
+	if (!is_unique_id(item->parent->server, item->id)) {
+		wlr_log(WLR_ERROR, "duplicate id '%s'; abort pipemenu", item->id);
+		return;
+	}
+
+	int pipe_fd = 0;
+	pid_t pid = spawn_piped(item->execute, &pipe_fd);
+	if (pid <= 0) {
+		wlr_log(WLR_ERROR, "Failed to spawn pipe menu process %s", item->execute);
+		return;
+	}
+
+	struct pipe_context *ctx = znew(*ctx);
+	ctx->server = item->parent->server;
+	ctx->item = item;
+	ctx->pid = pid;
+	ctx->pipe_fd = pipe_fd;
+	buf_init(&ctx->buf);
+
+	ctx->event_read = wl_event_loop_add_fd(ctx->server->wl_event_loop,
+		pipe_fd, WL_EVENT_READABLE, handle_pipemenu_readable, ctx);
+
+	ctx->event_timeout = wl_event_loop_add_timer(ctx->server->wl_event_loop,
+		handle_pipemenu_timeout, ctx);
+	wl_event_source_timer_update(ctx->event_timeout, PIPEMENU_TIMEOUT_IN_MS);
+}
+
 static void
 menu_process_item_selection(struct menuitem *item)
 {
 	assert(item);
 
+	/* Do not keep selecting the same item */
+	static struct menuitem *last;
+	if (item == last) {
+		return;
+	}
+	last = item;
+
 	if (!item->selectable) {
 		return;
 	}
 
-	/* We are on an item that has new mouse-focus */
+	/* We are on an item that has new focus */
 	menu_set_selection(item->parent, item);
 	if (item->parent->selection.menu) {
 		/* Close old submenu tree */
 		menu_close(item->parent->selection.menu);
 	}
 
+	/* Pipemenu */
+	if (item->execute) {
+		parse_pipemenu(item);
```

Comment:
I don't think the following code should be here as we didn't do anything at this point. E.g. we only spawned the subprocess and waiting for its stdout to return something later.

Reply:
removed via the latest patch.